File: MyApp/Models/Ballot/BallotFactory.cs
using System.Collections.Generic;

namespace MyApp.Models.Ballot
{
    public class BallotFactory
    {
        public static IBallot CreateBallot(
            BallotType type,
            int id,
            string title,
            string description,
            List<Candidate> candidates)
        {
            return type switch
            {
                BallotType.YesNo => new YesNoBallot
                {
                    Id = id,
                    Title = title,
                    Description = description,
                    Candidates = candidates
                },
                BallotType.RankedChoice => new RankedChoiceBallot
                {
                    Id = id,
                    Title = title,
                    Description = description,
                    Candidates = candidates
                },
                BallotType.MultipleChoice => new MultipleChoiceBallot
                {
                    Id = id,
                    Title = title,
                    Description = description,
                    Candidates = candidates
                },
                _ => throw new System.ArgumentException($"Unsupported ballot type: {type}")
            };
        }
    }
} 

File: client/src/services/ballot/BallotFactory.ts
import { IBallot, BallotType } from './IBallot';
import { YesNoBallot } from './YesNoBallot';
import { RankedChoiceBallot } from './RankedChoiceBallot';
import { MultipleChoiceBallot } from './MultipleChoiceBallot';
import { Candidate } from '../candidateService';

export class BallotFactory {
    static createBallot(
        type: BallotType,
        id: number,
        title: string,
        description: string,
        candidates: Candidate[]
    ): IBallot {
        switch (type) {
            case BallotType.YesNo:
                return new YesNoBallot(id, title, description, candidates);
            case BallotType.RankedChoice:
                return new RankedChoiceBallot(id, title, description, candidates);
            case BallotType.MultipleChoice:
                return new MultipleChoiceBallot(id, title, description, candidates);
            default:
                throw new Error(`Unsupported ballot type: ${type}`);
        }
    }
} 